File: problems/Letter Combinations of a Phone Number/letterCombinations.cs
public class Solution {
    public IList<string> LetterCombinations(string digits) {
        var map = new Dictionary<char, string>() { {'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};
        var result = new List<string>();

        if (0 < digits.Length) {
            backtrack(result, map, digits, new StringBuilder());
        }

        return result;
    }

    private void backtrack(IList<string> res, Dictionary<char, string> map, string digits, StringBuilder sb, int idx = 0) {
        if (digits.Length == sb.Length) {
            res.Add(sb.ToString());
        } else {
            var digit = digits[idx];
            var letters = map[digit];

            foreach (var letter in letters) {
                sb.Append(letter);
                backtrack(res, map, digits, sb, 1 + idx);
                sb.Remove(sb.Length - 1, 1);
            }
        }
    }
}

File: problems/Letter Combinations of a Phone Number/letterCombinations.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        HashMap<Character, String> digitToLettersMap = new HashMap<>();
        
        digitToLettersMap.put('2', "abc");
        digitToLettersMap.put('3', "def");
        digitToLettersMap.put('4', "ghi");
        digitToLettersMap.put('5', "jkl");
        digitToLettersMap.put('6', "mno");
        digitToLettersMap.put('7', "pqrs");
        digitToLettersMap.put('8', "tuv");
        digitToLettersMap.put('9', "wxyz");
        
        backtrack(digits, digitToLettersMap, 0, result, new StringBuilder());

        return result;
    }

    private void backtrack(String digits, HashMap<Character, String> digitToLettersMap, int idx, List<String> result, StringBuilder path) {
        if (digits.length() == path.length()) {
            if (0 < path.length()) {
                result.add(path.toString());
            }
            
            return;
        }

        String letters = digitToLettersMap.get(digits.charAt(idx));

        for (char letter : letters.toCharArray()) {
            path.append(letter);
            backtrack(digits, digitToLettersMap, 1 + idx, result, path);
            path.deleteCharAt(path.length() - 1);
        }
    }
}
